Initial Stubs for Compression Streams API

stwrt opened the pull request:
#### 9e3619bb6e46fb185a5f4e87579b0577488ba583
<pre>
Initial Stubs for Compression Streams API
<a href="https://bugs.webkit.org/show_bug.cgi?id=241964">https://bugs.webkit.org/show_bug.cgi?id=241964</a>

Reviewed by Alex Christensen and Brent Fulgham.

Implement initial stubs for Compression Streams API.

* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/compression/CompressionStream.cpp: Added.
(WebCore::CompressionStream::createStreams):
(WebCore::CompressionStream::readable):
(WebCore::CompressionStream::writable):
* Source/WebCore/Modules/compression/CompressionStream.h: Added.
(WebCore::CompressionStream::~CompressionStream):
(WebCore::CompressionStream::create):
(WebCore::CompressionStream::getCompressionFormat):
(WebCore::CompressionStream::CompressionStream):
* Source/WebCore/Modules/compression/CompressionStream.idl: Added.
* Source/WebCore/Modules/compression/DecompressionStream.cpp: Added.
(WebCore::DecompressionStream::createStreams):
(WebCore::DecompressionStream::readable):
(WebCore::DecompressionStream::writable):
* Source/WebCore/Modules/compression/DecompressionStream.h: Added.
(WebCore::DecompressionStream::~DecompressionStream):
(WebCore::DecompressionStream::create):
(WebCore::DecompressionStream::getCompressionFormat):
(WebCore::DecompressionStream::DecompressionStream):
* Source/WebCore/Modules/compression/DecompressionStream.idl: Added.
* Source/WebCore/Modules/compression/Formats.h: Added.
* Source/WebCore/Modules/streams/GenericTransformStream.idl: Added.
* Source/WebCore/Modules/websockets/WebSocket.cpp:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventTargetFactory.in:
* Source/WebCore/features.json:

Canonical link: <a href="https://commits.webkit.org/252071@main">https://commits.webkit.org/252071@main</a>
</pre>

brentfulgham commented on Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
```diff
@@ -457,6 +457,18 @@ CanvasColorSpaceEnabled:
+  humanReadableDescription: "Enable Compression Stream API"
+  defaultValue:
+    WebKitLegacy:
+      default: true
```
We tend to avoid adding new features to WebKitLegacy to discourage its use, and to avoid triggering bugs that we will have to investigate on a deprecated platform.

brentfulgham replied:
We also usually start with the setting set to `false` until we are confident the feature is ready to ship.

I would suggest leaving all of these 'false' for now, but make sure to enable the features in WKTR.

brentfulgham commented on Source/WebCore/Modules/compression/CompressionStream.h:
```diff
@@ -0,0 +1,124 @@
+    enum CompressionFormat { gzip, deflate, deflate_raw };
+
+    CompressionStream(Document& document, String string, CompressionFormat format) : ActiveDOMObject(document), m_string(string), m_format(format) {
+        
```
Nit: I don't think we need the blank lines around the small implementation here.

brentfulgham commented on Source/WebCore/Modules/compression/CompressionStream.h:
```diff
@@ -0,0 +1,124 @@
+    ~CompressionStream() {}
+
+    void ref() { }
+    void deref() { }
```
I'm surprised you have these dummy ref/deref commands, fi you are bothering to make it refcounted. Are these stubs until you hook up an underlying connection to Compression or something?

stwrt replied:
I was seeing some build failures initially without these definitions. Was planning on looking into this more.

stwrt replied:
These should be removed in latest commit.

brentfulgham commented on Source/WebCore/Modules/websockets/WebSocket.cpp:
```diff
@@ -49,6 +49,7 @@
 #include "MessageEvent.h"
 #include "MixedContentChecker.h"
 #include "ResourceLoadObserver.h"
+#include "RuntimeEnabledFeatures.h"
```
This file was removed in a very recent check-in.

stwrt replied:
This should be removed in my newest commit.

brentfulgham reviewed (changes requested):
This looks okay as a reasonable first step, but I'd suggest you adjust the commit message to indicate that this is "Initial stubs for Compression Streams API" (or something).

Can you also update `Source/WebCore/features.json` with a JSON chunk that indicates that provides a `"status": "In Development"` for this API?

brentfulgham reviewed (changes requested):
I also notice that this appears to be two separate checkins. Can you please merge them into a single commit?

Also: Please run the `check-webkit-style` on your change and fix before landing,

brentfulgham commented on Source/WebCore/Modules/compression/CompressionStream.h:
```diff
@@ -0,0 +1,119 @@
+
+namespace WebCore {
+
+class CompressionStream : public RefCounted<CompressionStream>,
```
Is CompressionStream not a GenericTransformStream? Should they have some kind of relationship?

stwrt replied:
The CompressionStream includes the GenericTransfromStream in the IDL.

It didn't look like we had a GenericTransformStream implementation before. I might expand upon it when adding the DecompressionStream API.

achristensen07 replied:
You should add DecompressionStream stubs at the same time as this.  They are a pair.

brentfulgham commented on Source/WebCore/Modules/streams/GenericTransformStream.h:
```diff
@@ -0,0 +1,26 @@
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
```
Strange to have a completely empty file -- not even a default GenericTransformStream struct. Do we need this (or perhaps this will be expanded further in a follow-up?)

stwrt replied:
Let me delete this. I didn't have any plans on using this immediately.

brentfulgham reviewed (approved):
Looks like a good first step!

achristensen07 commented on Source/WebCore/Modules/compression/CompressionStream.cpp:
```diff
@@ -0,0 +1,95 @@
+    auto simpleWritableStreamSink = SimpleWritableStreamSink::create([readableStreamSource = m_readableStreamSource, weakThis = WeakPtr { *this }](auto& context, auto value) -> ExceptionOr<void> {
+        if (!context.globalObject())
+            return Exception { InvalidStateError };
+        // auto& globalObject = *JSC::jsCast<JSDOMGlobalObject*>(context.globalObject());
```
Let's not put commented code into our repo.

achristensen07 commented on Source/WebCore/Modules/compression/CompressionStream.cpp:
```diff
@@ -0,0 +1,95 @@
+        // auto scope = DECLARE_THROW_SCOPE(globalObject.vm());
+        
+        // TODO: remove (making the compiler happy)
+        if (value) { }
```
Ditto.  Remove.

brentfulgham replied:
For now you could do `UNUSED_PARAM(value);`

achristensen07 commented on Source/WebCore/Modules/compression/CompressionStream.h:
```diff
@@ -0,0 +1,121 @@
+    ~CompressionStream() { }
+
+    // TODO: Cleanup
+    void ref() { }
```
Nope, don't put this in the repo.

brentfulgham commented on Source/WebCore/Modules/compression/CompressionStream.h:
```diff
@@ -0,0 +1,124 @@
+    // deflate      "ZLIB Compressed Data Format"   [RFC1950]
+    // deflate-raw  "The DEFLATE algorithm"         [RFC1951]
+    enum CompressionFormat { Gzip, Deflate, Deflate_raw };
+    CompressionStream(Document& document, String string, CompressionFormat format) 
```
This constructor should probably be private (the approach we use in other DOM Classes). The '`create`' method is public.

brentfulgham commented on Source/WebCore/Modules/compression/CompressionStream.h:
```diff
@@ -0,0 +1,124 @@
+    // Compression Formats
+    // gzip         "GZIP file format"              [RFC1952]
+    // deflate      "ZLIB Compressed Data Format"   [RFC1950]
+    // deflate-raw  "The DEFLATE algorithm"         [RFC1951]
```
We probably don't need these individual comments -- they could go in the Commit message, but not needed elsewhere.

stwrt replied:
I added these because it was just nice to view at a quick glance what these formats referred to, but I will remove them.

brentfulgham commented on Source/WebCore/Modules/compression/CompressionStream.h:
```diff
@@ -0,0 +1,124 @@
+    ~CompressionStream() { }
+
+    void ref() { m_refCount++; }
+    void deref() { m_refCount--; }
```
Maybe just:
`using RefCounted<CompressionStream>::deref;`

brentfulgham commented on Source/WebCore/Modules/compression/CompressionStream.h:
```diff
@@ -0,0 +1,124 @@
+private:
+
+    // Ref Counting
+    unsigned m_refCount { 1 };
```
Then I think you can remove this (which is hiding the m_refCount from the RefCounted base class).

brentfulgham commented on Source/WebCore/Modules/compression/CompressionStream.h:
```diff
@@ -0,0 +1,124 @@
+
+    ~CompressionStream() { }
+
+    void ref() { m_refCount++; }
```
`using RefCounted<CompressionStream>::ref;`

brentfulgham commented on Source/WebCore/Modules/compression/DecompressionStream.h:
```diff
@@ -0,0 +1,124 @@
+    // deflate      "ZLIB Compressed Data Format"   [RFC1950]
+    // deflate-raw  "The DEFLATE algorithm"         [RFC1951]
+    enum CompressionFormat { Gzip, Deflate, Deflate_raw };
+    DecompressionStream(Document& document, String string, CompressionFormat format) 
```
This constructor should be private.

brentfulgham commented on Source/WebCore/Modules/compression/DecompressionStream.h:
```diff
@@ -0,0 +1,124 @@
+    // Compression Formats
+    // gzip         "GZIP file format"              [RFC1952]
+    // deflate      "ZLIB Compressed Data Format"   [RFC1950]
+    // deflate-raw  "The DEFLATE algorithm"         [RFC1951]
```
These comments aren't needed.

brentfulgham commented on Source/WebCore/Modules/compression/DecompressionStream.h:
```diff
@@ -0,0 +1,124 @@
+
+    ~DecompressionStream() { }
+
+    void ref() { m_refCount++; }
```
`using RefCounted<DecompressionStream>::ref;`

brentfulgham commented on Source/WebCore/Modules/compression/DecompressionStream.h:
```diff
@@ -0,0 +1,124 @@
+    ~DecompressionStream() { }
+
+    void ref() { m_refCount++; }
+    void deref() { m_refCount--; }
```
`using RefCounted<DecompressionStream>::deref;`

brentfulgham commented on Source/WebCore/Modules/compression/DecompressionStream.h:
```diff
@@ -0,0 +1,124 @@
+private:
+
+    // Ref Counting
+    unsigned m_refCount { 1 };
```
You shouldn't need this -- it's overriding the one in the base class.

brentfulgham commented on Source/WebCore/Modules/compression/DecompressionStream.h:
```diff
@@ -0,0 +1,124 @@
+    // gzip         "GZIP file format"              [RFC1952]
+    // deflate      "ZLIB Compressed Data Format"   [RFC1950]
+    // deflate-raw  "The DEFLATE algorithm"         [RFC1951]
+    enum CompressionFormat { Gzip, Deflate, Deflate_raw };
```
This implies we will sometimes do:
`DecompressionStream::CompressionFormat::Gzip` and other times `CompressionStream::CompressionFormat::Gzip`, depending on the call site.

It seems like having a single enum for both cases would be better. But I'm not sure where it would be best to declare that enum -- perhaps in a file called "Modules/compression/formats.h" or something?

This should probably also be:
enum class CompressionFormat : uint8_t { Gzip, Deflate, Deflate_raw };

stwrt replied:
Yes, I was thinking about doing this earlier. I wasn't sure if I just wanted to create an entire file just for one shared enum. I'll update the patch though with this new file.

brentfulgham reviewed (changes requested):
Sorry -- I noticed a few more things I think we should fix before landing. But I think this looks really good! It's very close!

achristensen07 commented on Source/WebCore/Modules/compression/CompressionStream.h:
```diff
@@ -0,0 +1,116 @@
+        if (string == String("deflate-raw", 11))
+            return adoptRef(*new CompressionStream(document, string, Formats::Deflate_raw));
+
+        // Section 5
```
Link to spec?

stwrt replied:
I would prefer to add this in a follow up patch. I was planning to add the spec towards the top of the file. I will probably be adding more references to the spec, and don't necessarily want to litter it everywhere throughout the file.

achristensen07 commented on Source/WebCore/Modules/compression/CompressionStream.cpp:
```diff
@@ -0,0 +1,92 @@
+    m_readable = readable.releaseReturnValue();
+
+    // setup writable stream
+    auto simpleWritableStreamSink = SimpleWritableStreamSink::create([readableStreamSource = m_readableStreamSource, weakThis = WeakPtr { *this }](auto& context, auto value) -> ExceptionOr<void> {
```
This is a lot of duplicate code.  Maybe make a common superclass and put the code there?  Maybe it could be called GenericTransformStream and be RefCounted.

stwrt replied:
I agree there is a lot of duplicate code at the moment. As the implementation progresses I was planning on seeing exactly how to best abstract it.

stwrt committed:
Initial Stubs for Compression Streams API
https://bugs.webkit.org/show_bug.cgi?id=241964

Reviewed by Alex Christensen and Brent Fulgham.

Implement initial stubs for Compression Streams API.

* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/compression/CompressionStream.cpp: Added.
(WebCore::CompressionStream::createStreams):
(WebCore::CompressionStream::readable):
(WebCore::CompressionStream::writable):
* Source/WebCore/Modules/compression/CompressionStream.h: Added.
(WebCore::CompressionStream::~CompressionStream):
(WebCore::CompressionStream::create):
(WebCore::CompressionStream::getCompressionFormat):
(WebCore::CompressionStream::CompressionStream):
* Source/WebCore/Modules/compression/CompressionStream.idl: Added.
* Source/WebCore/Modules/compression/DecompressionStream.cpp: Added.
(WebCore::DecompressionStream::createStreams):
(WebCore::DecompressionStream::readable):
(WebCore::DecompressionStream::writable):
* Source/WebCore/Modules/compression/DecompressionStream.h: Added.
(WebCore::DecompressionStream::~DecompressionStream):
(WebCore::DecompressionStream::create):
(WebCore::DecompressionStream::getCompressionFormat):
(WebCore::DecompressionStream::DecompressionStream):
* Source/WebCore/Modules/compression/DecompressionStream.idl: Added.
* Source/WebCore/Modules/compression/Formats.h: Added.
* Source/WebCore/Modules/streams/GenericTransformStream.idl: Added.
* Source/WebCore/Modules/websockets/WebSocket.cpp:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventTargetFactory.in:
* Source/WebCore/features.json:

Canonical link: https://commits.webkit.org/252071@main
